### PR TITLE
test(semantic): test harness

### PR DESCRIPTION
**Pull request by DonIsaac**
A test harness for checking results of semantic analysis.

I got tired of writing ad-hoc test cases when finding bugs in semantic analysis, so I made this.

### CODE REVIEW

File: crates/oxc_semantic/Cargo.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ oxc_ast         = { workspace = true }
 oxc_syntax      = { workspace = true }
 oxc_diagnostics = { workspace = true }
 oxc_index       = { workspace = true }
+miette          = { workspace = true, features = ["fancy-no-backtrace"] }
```
Comment (Boshen) [inline crates/oxc_semantic/Cargo.toml:20]: should be dev dependency?

**Comment by Boshen** [comment]
I took a look at how Rome does it, https://github.com/rome/tools/tree/main/crates/rome_js_semantic/src/tests, I like your approach more.

**Comment by Boshen** [comment]
This is getting big, shall we merge the basic setup before adding more details?

**Comment by DonIsaac** [comment]
Sure, I'll save further changes for following PRs. I think this is ready for review.

File: crates/oxc_semantic/tests/util/mod.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+extern crate miette;
+use oxc_semantic::{Reference, Semantic, SemanticBuilder, SymbolFlags, SymbolId};
+use oxc_span::{Atom, SourceType};
+
+pub struct SemanticTester {
+    allocator: Allocator,
+    source_type: SourceType,
+    source_text: &'static str,
+    /// SemanticBuilder option
+    module_builder: bool,
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:18]: Maybe `use_module_record_builder`? Let's try to be as explicit as possible.

File: crates/oxc_semantic/tests/util/mod.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            .collect()
+    }
+}
+
+pub struct SymbolTester<'a> {
+    parent: &'a SemanticTester,
+    /// Reference to semantic analysis results, from [`SemanticTester`]
+    semantic: Semantic<'a>,
+    /// Name of the subject symbol
+    target: String,
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:144]: `target` is too vague, maybe `target_symbol_name`?

File: crates/oxc_semantic/tests/util/mod.rs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    /// Test case source code.
+    ///
+    /// Technically, `semantic` also has a reference to the source code with lifetime 'a and this
+    /// _could_ also be 'a, but we need it to be 'static to work with miette.
+    // source: &'static str,
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:149]: If it's for tests, should be fine with `Arc<str>`

File: crates/oxc_semantic/tests/util/mod.rs [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    /// Symbol data, or error if not found
+    data: Result<SymbolId, oxc_diagnostics::Error>,
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:151]: `data` is too vague, maybe `test_symbol_result`?

File: crates/oxc_semantic/tests/util/mod.rs [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                        found_flags
+                    ))
+                }
+            }
+            err => err,
+        };
+        self
+    }
+
+    pub fn has_number_of_reads(self, ref_count: usize) -> Self {
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:227]: nitpick on the API: what about `assert_xxx` and use `&self`? Chaining these methods calls are equivalent of calling them individually. Calling them individually has the advance of writing test cases where if statements are required. e.g.

```
let tester = Tester::new();
if is_condition {
   tester.assert_xxx();
}
```

File: crates/oxc_semantic/tests/util/mod.rs [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    }
+}
+
+impl<'a> From<SymbolTester<'a>> for Result<(), Error> {
+    fn from(val: SymbolTester<'a>) -> Self {
+        val.data.map(|_| {}).map_err(|e| e.with_source_code(val.parent.source_text))
+    }
+}
+
+// fn graphic_hook(diag: dyn Diagnostic) -> GraphicalReportHandler {
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:298]: Remove?

**Review by Boshen — CHANGES_REQUESTED** [review]
Thank you for working on this. Overall this is really good, and is something we really need.

I commented with a lot of nitpicks because I want us to align on the same coding style.

File: crates/oxc_semantic/tests/modules.rs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        export { foo };
+        export default defaultExport;
+        ",
+    )
+    .with_module_record_builder(true);
+
+    test.has_some_symbol("foo").is_exported().test();
+
+    // FIXME: failing
+    // test.has_some_symbol("defaultExport").is_exported().test();
```
[the remark this answers is not in the document]
Comment (DonIsaac) [reply crates/oxc_semantic/tests/modules.rs:32]: I'm confused - why wouldn't the list of exported bindings contain the default export? is this part of the spec?

File: crates/oxc_semantic/tests/util/mod.rs (hunk 6 above)
Comment (DonIsaac) [reply crates/oxc_semantic/tests/util/mod.rs:227]: I personally like the chaining, it builds a sentence. I agree with your first point, how about something like `should_have_read_count`? To me, this sounds more like a predicate than `assert_xxx`

File: crates/oxc_semantic/tests/modules.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Boshen) [reply crates/oxc_semantic/tests/modules.rs:32]: Yeah, take a look at it `ExportExportName` in `ModuleRecord`, it's subtle but there is a difference. Our implementation conforms to the spec https://tc39.es/ecma262/#exportentry-record

I have plan to test import and export here https://github.com/web-infra-dev/oxc/issues/623

File: crates/oxc_semantic/tests/util/mod.rs (hunk 6 above)
Comment (Boshen) [reply crates/oxc_semantic/tests/util/mod.rs:226]: Ok, `should_have_read_count` or `has_number_of_reads` are both fine.

**Review by Boshen — APPROVED** [review]

**Pull request merged**